index.js

```diff
@@ -1,469 +1,173 @@
+    this.userAgent = `Tallahassee/${version}`;
+    this.protocol = `${originRequestHeaders["x-forwarded-proto"] || "http"}:`;
+    this.referrer = originRequestHeaders.referer;
+  }
+  async load(uri, headers, statusCode = 200) {
+    const requestHeaders = normalizeHeaders(headers);
+    if (requestHeaders["user-agent"]) this.userAgent = requestHeaders["user-agent"];
 
-        agent.jar.setCookies(requestHeaders.cookie.split(";").map((c) => c.trim()).filter(Boolean), cookieDomain, "/", isSecure);
-      }
+    if (requestHeaders.cookie) {
+      const publicHost = getLocationHost(requestHeaders);
+      const parsedUri = url.parse(uri);
+      const cookieDomain = parsedUri.hostname || publicHost || this.originHost || "127.0.0.1";
+      const isSecure = (parsedUri.protocol || this.protocol) === "https:";
 
-      return fetch(uri, {
-        method: "GET",
-        headers: requestHeaders,
-      }).then((resp) => {
-        assert.equal(resp.status, statusCode, `Unexepected status code. Expected: ${statusCode}. Actual: ${resp.statusCode}`);
-        assert(resp.headers.get("content-type").match(/text\/html/i), `Unexepected content type. Expected: text/html. Actual: ${resp.headers["content-type"]}`);
-        return resp;
-      }).then(loadDOMContent);
+      this.agent.jar.setCookies(requestHeaders.cookie.split(";").map((c) => c.trim()).filter(Boolean), cookieDomain, "/", isSecure);
     }
 
-    function fetch(uri, requestOptions = {}) {
-      let numRedirects = 0;
-      return makeRequest(uri, requestOptions).then(handleResponse);
-
-      function handleResponse(res) {
-        const setCookieHeader = res.headers.get("set-cookie");
-        if (setCookieHeader) {
-          const cookieDomain = url.parse(res.url).hostname;
-          saveToJar(setCookieHeader, cookieDomain);
-        }
-
-        if (res.status > 300 && res.status < 309 && requestOptions.redirect !== "manual") {
-          numRedirects++;
-          if (numRedirects > 20) {
-            throw new Error("Too many redirects");
-          }
-          const location = res.headers.get("location");
-          const redirectOptions = {...requestOptions};
-
-          if (res.status === 307 || res.status === 308) {
-            // NO-OP
-          } else {
-            redirectOptions.method = "GET";
-            redirectOptions.body = undefined;
-          }
-
-          return makeRequest(location, redirectOptions).then(handleResponse);
-        }
-
-        return res;
-      }
+    const resp = await this.fetch(uri, {
+      method: "GET",
+      headers: requestHeaders,
+    });
+    assert.equal(resp.status, statusCode, `Unexepected status code. Expected: ${statusCode}. Actual: ${resp.statusCode}`);
+    assert(resp.headers.get("content-type").match(/text\/html/i), `Unexepected content type. Expected: text/html. Actual: ${resp.headers["content-type"]}`);
+    const browser = new BrowserTab(this, resp);
+    return browser.load();
```

Comment:
Feels weird that a `WebPage` instance creates a `browser`. Should be reversed?
Also that `browser` comes from the `BrowserTab` class?